### PR TITLE
RESTEASY-3006 Allow microprofile-client to inject Vertx engine for HTTP/2 communication

### LINKED ISSUE #8: Allow microprofile-client to inject Vertx engine for HTTP/2 communication
Replaces https://issues.redhat.com/browse/RESTEASY-3006

### PR DESCRIPTION
**Pull request by liweinan**
resolves #8

### CODE REVIEW

File: testsuite/integration-tests/pom.xml [hunk 1]
```diff
@@ -68,6 +68,11 @@
             <artifactId>smallrye-config</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.jboss.resteasy</groupId>
+            <artifactId>resteasy-client-vertx</artifactId>
+            <version>${version.org.jboss.resteasy}</version>
+        </dependency>
```
Comment (jamezp) [inline testsuite/integration-tests/pom.xml:75]: This should be defined in the resteasy-microprofile-test-bom and have a `<scope>test</scope>`

**Review by jamezp — CHANGES_REQUESTED** [review]
Overall I think this looks okay. I want to think about if we can get it to work without having to use the `RestClientBuilderImpl` directly. One thing that comes to mind is doing something like:

```
HttpClientOptions options = new HttpClientOptions();
options.setSsl(true);
options.setProtocolVersion(HttpVersion.HTTP_2);
options.setUseAlpn(true);

builder.setProperty("org.jboss.resteasy.http.client.engine", new VertxClientHttpEngine(Vertx.vertx(), options));
```
Then the `RestClientBuilderImpl` itself looks for the property. It requires knowledge of the `ClientHttpEngine`, but that seems okay as that is a public API and IMO the `RestClientBuilderImpl` should not be a public API.

File: testsuite/integration-tests/pom.xml (hunk 1 above)
Comment (liweinan) [reply testsuite/integration-tests/pom.xml:75]: @jamezp Okay! I'll work on it.

**Comment by liweinan** [comment]
@jamezp Unfortunately `RestClientBuilder` doesn't have a `setProperty()` method and it's defined by eclipse community:

image

Shall I submit a request to add a method into the interface to the eclipse community? Is there any workflow to submit it?

**Comment by jamezp** [comment]
> @jamezp Unfortunately `RestClientBuilder` doesn't have a `setProperty()` method and it's defined by eclipse community:
> 
> image
> 
> Shall I submit a request to add a method into the interface to the eclipse community? Is there any workflow to submit it?

There should be as it extends `javax.ws.rs.core.Configurable`. I had a typo in my example and it should be `builder.property("org.jboss.resteasy.http.client.engine", new VertxClientHttpEngine(Vertx.vertx(), options));`. The `builder.getConfiguration().getProperty("org.jboss.resteasy.http.client.engine")` could be used in the implementation to lookup the requested engine.

**Comment by liweinan** [comment]
@jamezp Thanks for the suggestion! I'll go on working on this :D

**Comment by liweinan** [comment]
@jamezp I have modified all the places you've mentioned, could you please help to do a final check? Thanks! :D

(the test error in above I checked is because of Github CI platform network problem)

After this PR is merged I'll:

- Write a blog related with HTTP/2 client support on `resteasy` side and `resteasy-microprofile` side.
- Create a similar PR to `resteasy-reactive-client` (to Quarkus).

**Comment by liweinan** [comment]
@jamezp I merged the PR, if there is any problem pls let me know.

**Pull request merged**